File: Backend/Pricing.Application/Domain/Entities/Cupom.cs
using Baldan.Pricing.Application.Domain.Enums;
using Baldan.Pricing.Application.Models.Enums;
using Mercado.Craibas.Application.Domain.Entities;

namespace Baldan.Pricing.Application.Domain.Entities
{
    public class Cupom : EntityBase
    {
        public string Name_Cupom { get; set; }
        public string Cod_Cupom { get; set; }
        public string Description { get; set; }
        public double Discount { get; set; }
        public DiscountType Discount_Type { get; set; }
        public bool Active { get; set; }
        public bool Show_Flash_Offer { get; set; }
        public double? Minimum_Value { get; set; }
        public double? Maximum_Discount { get; set; }
        public int? Quantity_Uses { get; set; }
        public int? Quantity_Used { get; set; }
        public int? Per_User_Limit { get; set; }
        public bool First_Order_Only { get; set; }
        public DateTime? Date_Start { get; set; }
        public DateTime? Date_End { get; set; }
        public ICollection<Orders> Orders { get; set; } = [];

        public ICollection<Coupon_Product> Coupon_Products { get; set; } = [];

        public ICollection<Coupon_Category> Coupon_Categories { get; set; } = [];

        public ICollection<Coupon_Use> Coupon_Uses { get; set; } = [];
    }
}

File: src/models/Cupom.ts
export interface Cupom {
    id?: number;
    name_Cupom?: string;
    cod_Cupom?: string;
    description?: string;
    discount_Type?: number;
    discont?: number;
    active?: boolean;
    minimum_Value?: number | null;
    maximum_Discount?: number | null;
    quantity_Uses?: number | null;
    quantity_Used?: number | null;
    per_User_Limit?: number | null;
    first_Order_Only?: boolean;
    application?: string;
    date_Start?: string | null;
    date_End?: string | null;
    date_end?: string | null;
    productIds?: number[];
    categoryIds?: number[];
    show_Flash_Offer: boolean;
}

export interface CupomAdmin {
    id?: number;
    name_Cupom?: string;
    cod_Cupom?: string;
    description?: string;
    discount_Type?: DiscountType;
    discount?: number;
    active?: boolean;
    minimum_Value?: number | null;
    maximum_Discount?: number | null;
    quantity_Uses?: number | null;
    quantity_Used?: number | null;
    per_User_Limit?: number | null;
    first_Order_Only?: boolean;
    application?: string;
    date_Start?: string | null;
    date_End?: string | null;
    productIds?: Coupon_Product[];
    categoryIds?: Coupon_Category[];
    // couponUse?: Coupon_Use[];
    show_Flash_Offer: boolean;
}
export type UpdateCouponRequest = CupomAdmin & {
    addedProducts: number[];
    removedProducts: number[];

    addedCategories: number[];
    removedCategories: number[];
};
export interface Coupon_Category {
    id: number;
    id_Cupom: number;
    id_Category: number;
    insertDate?: Date;
    updateDate?: Date | null;
}

export interface Coupon_Product {
    id: number;
    id_Cupom: number;
    id_Product: number;
    insertDate?: Date;
    updateDate?: Date;
}
export interface Coupon_Use {
    id: number;
    id_Cupom: number;
    Id_Order: number;
    Id_User: number;
    insertDate: Date;
    updateDate: Date;
}
export enum DiscountType {
    Percentage = "Percentage",
    FixedValue = "FixedValue",
    FreeShipping = "FreeShipping"
}
